{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe design workflow\n",
    "## Inputs for DesignTargets\n",
    "1. Reference fasta file\n",
    "    - Sequence names must be the be simple name\n",
    "    - Must contain all the sequences one wishes to screen against off target effects\n",
    "2. (optional) List of sequence names that are not expressed\n",
    "\n",
    "## Inputs for GeneExpression\n",
    "1. Expression FPKM or counts for different genes (must have same names as used in DesignTargets\n",
    "2. GTF file\n",
    "\n",
    "# Dependencies\n",
    "1. OligoArray2.jar\n",
    "2. formatdb (legacy)\n",
    "3. BLAST (legacy)\n",
    "4. Bio python package\n",
    "5. pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "0.2 Update Changes\n",
    "\n",
    "Change parse oligo to allow user to specify path in order to reload already completed stuffs.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from tempfile import mkdtemp\n",
    "from os import rmdir, path, mkdir, system\n",
    "from shutil import rmtree\n",
    "from subprocess import call\n",
    "import uuid\n",
    "from pandas import DataFrame\n",
    "\n",
    "\n",
    "# Path to formatdb, BLAST, and any other dependent tools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class DesignTargets():\n",
    "    \"\"\"\n",
    "    Class designed to generate hybridization probes against target mRNA while \n",
    "    minimizing the offtarget binding of the probe.\n",
    "    Inputs:\n",
    "    Ensemble ID of targets and a fasta of all off target references to screen against.\n",
    "    OligoArray2.1\n",
    "    Old BLAST\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 transcript_fasta = '/data/v24.transcripts.trimmed.fa', \n",
    "                 temp_dir = '/data/'):\n",
    "        \"\"\"\n",
    "        Initiliaze the class with a fasta file of offtarget sequences.\n",
    "        A root directory to store temporary files.\n",
    "        Some parameters of the probe selection algorithms.\n",
    "        \"\"\"\n",
    "        self.jarpth = '/data/wollman/tools/bin/OligoArray2.jar'\n",
    "        self._assignReferences(transcript_fasta)\n",
    "        self._temp_dir = path.join(temp_dir, str(uuid.uuid4()))\n",
    "        mkdir(self._temp_dir)\n",
    "        self.gc_min = 20\n",
    "        self.gc_max = 80\n",
    "        self.length = 30\n",
    "        self.min_t = 70\n",
    "        self.max_t = 100\n",
    "        self.xhybe = 72\n",
    "        self.tsecondary = 76\n",
    "        self.nthreads = 16\n",
    "        \n",
    "    def designProbes(self, transcript_id = None, not_expressed = None):\n",
    "        # Type checking inputs and property states\n",
    "        if transcript_id is None:\n",
    "            if self._targetTranscripts is None:\n",
    "                raise Exception('No known transcripts to design probes for.')\n",
    "            else:\n",
    "                transcript_id = self._targetTranscripts\n",
    "        elif not isinstance(transcript_id, list):\n",
    "            if isinstance(transcript_id, str):\n",
    "                transcript_id = [transcript_id]\n",
    "        if not_expressed is None:\n",
    "            blastdb = self._makeblastdb(fdir = path.join(self._temp_dir, 'db.fa'))\n",
    "        else:\n",
    "            trimmed_records = list(set(self._records.keys()) - set(not_expressed))\n",
    "            blastdb = self._makeblastdb(records=trimmed_records, fdir = path.join(self._temp_dir, 'db.fa'))\n",
    "        # Iterate over the transcripts that need targets designed\n",
    "        self._oligoarray2(transcript_id, blastdb=blastdb)\n",
    "                \n",
    "    def _oligoarray2(self, sequences, blastdb = None):\n",
    "        \"\"\"\n",
    "        Wrapper for OligoArray2.1 system call.\n",
    "        Given a list of sequence names run the oligoarray2.1 jar file.\n",
    "        Use the parameters as they are currently in the instance.\n",
    "        \"\"\"\n",
    "        fname = self._writefasta(sequences, fdir = path.join(self._temp_dir, 'temp.fa'))\n",
    "        oligos_out = path.join(self._temp_dir, 'oligos.txt')\n",
    "        log_out = path.join(self._temp_dir, 'oligos.log')\n",
    "        !java -Xmx4g -jar {self.jarpth} -i {fname} -d {blastdb} -o {oligos_out} -R {log_out} -n 196 -l 30 -L 30 -D 10000 -t {self.min_t} -T {self.max_t} -s {self.tsecondary} -x {self.xhybe} -N {self.nthreads} -g 30 -m \"GGGGGG;CCCCCC;TTTTTT;AAAAAA\" -p {self.gc_min} -P {self.gc_max}\n",
    "\n",
    "    def _assignReferences(self, fastafile):\n",
    "        \"\"\"\n",
    "        Assecory function to setup fasta files.\n",
    "        \"\"\"\n",
    "        if isinstance(fastafile, str):\n",
    "            self._records = SeqIO.index(fastafile, 'fasta')\n",
    "        elif isinstance(fastafile, SeqIO.index):\n",
    "            self._records = fastafile\n",
    "    def _writefasta(self, transcript_ids, temp=True, fdir = 'tempfile.fa'):\n",
    "        \"\"\"\n",
    "        Assecory function to write a fasta file containing the subset of \n",
    "        intended target sequence names.\n",
    "        \"\"\"\n",
    "        sequences = [self._records.get(seq) for seq in transcript_ids]\n",
    "        print(sequences)\n",
    "        with open(fdir, 'w') as handle:\n",
    "            SeqIO.write(sequences, handle, 'fasta')\n",
    "        return fdir\n",
    "    def _makeblastdb(self, records=None, fdir = 'tempdb.fa'):\n",
    "        \"\"\"\n",
    "        Transform the fasta file of possible off-target sequences \n",
    "        into a blastdb indexed format necessary for OligoArray2.1.\n",
    "        \"\"\"\n",
    "        if records is None:\n",
    "            records = self._records\n",
    "            with open(fdir, 'w') as handle:\n",
    "                SeqIO.write(records.values(), handle, 'fasta')\n",
    "        if isinstance(records, list):\n",
    "            with open(fdir, 'w') as handle:\n",
    "                SeqIO.write([self._records.get(i) for i in records], handle, 'fasta')\n",
    "        !formatdb -i {fdir} -o T -p F\n",
    "        return fdir\n",
    "    \n",
    "    def _parseOligoArrayOutput(self, fname='oligos.txt', fpth=None):\n",
    "        \"\"\"\n",
    "        Parse the output file into two dataframe objects.\n",
    "        One contains the parameters of the hybridixation between candidate probes \n",
    "        and their cognate mRNA. Another contains all the possible offtarget binding \n",
    "        sites between candidate probes and sequences in the off-target database.\n",
    "        The designer writes a CSV that this parses.\n",
    "        \"\"\"\n",
    "        probe_colnames = ['target', 'pos', 'len', 'dG', 'dH', 'dS', 'Tm', 'seq', 'off_df']\n",
    "        probes = []\n",
    "        offtargets = []\n",
    "        if fpth is not None:\n",
    "            filename = fpth\n",
    "        else:\n",
    "            filename = path.join(self._temp_dir, fname)\n",
    "        with filename as f:\n",
    "            for line in f.readlines():\n",
    "                line = line.rstrip()\n",
    "                columns = line.split('\\t')\n",
    "                seq = columns[-1]\n",
    "                probe_target = columns[0]\n",
    "                pos = columns[1]\n",
    "                offt = columns[-2].split(';')\n",
    "                parent = offt[0].strip()\n",
    "                # Edge case in the output formatting\n",
    "                if parent != probe_target:\n",
    "                    parent_split = parent.find(',')\n",
    "                    offt.append(parent[parent_split+1:])\n",
    "                    parent = probe_target\n",
    "                    print('Warning', parent, probe_target)\n",
    "                offtarg_sites = []\n",
    "                for off_site in offt[1:]:\n",
    "                    names, binding = off_site.split('(')\n",
    "                    names = names.strip()\n",
    "                    names = names.split(',')\n",
    "                    names = [n.strip() for n in names]\n",
    "                    binding = binding.strip(')').split(' ')\n",
    "                    for n in names:\n",
    "                        offtarg_sites.append([parent, pos, n]+binding)\n",
    "                probes.append(columns[:-2]+[seq]+\n",
    "                              [DataFrame(offtarg_sites, \n",
    "                                         columns=['target', 'pos', 'offtarget', \n",
    "                                                  'dG', 'dH', 'dS', 'tm', 'offseq'])\n",
    "                              ])\n",
    "                \n",
    "        probe_df = DataFrame(probes, columns=probe_colnames)\n",
    "        return probe_df\n",
    "    \n",
    "    def cleanup(self):\n",
    "        rmtree(self._temp_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class MerprobeOligos():\n",
    "    \n",
    "    def __init__(primer_path = '/data/wollman/hypefish/datafiles/PrimerSeqs.fasta'):\n",
    "        # Create a dictionary of possible primers each time it generates a primer pair \n",
    "        # they will be removed from the dict to prevent reuse\n",
    "        primer_seqs = {}\n",
    "        fasta = SeqIO.parse(primer_path, 'fasta')\n",
    "        for seq in fasta:\n",
    "            primer_seqs[seq.id] = str(seq.seq)\n",
    "        fasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "class ScreenTargets():\n",
    "    def __init__(self, df, arg=None):\n",
    "        self.probedf = df\n",
    "    def _parseProbeQuality(self, input_df, arg = None):\n",
    "        \"\"\"\n",
    "        Make a df with columns for each candidate probe \n",
    "        that ranks how good and specific it is.\n",
    "        \"\"\"\n",
    "        if arg is None:\n",
    "            arg = {'output': ['sum_exp', 'numoff', ]}\n",
    "    def addQualityColumns(self, exp_df = None):\n",
    "        \"\"\"\n",
    "        Notes: exp_df must have a column transcript_id and SUM_FPKM.\n",
    "        \"\"\"\n",
    "        self.probedf['numoff'] = [row.off_df.shape[0] for ix, row in self.probedf.iterrows()]\n",
    "        if exp_df is not None:\n",
    "            exp_values = []\n",
    "            for ix, row in self.probedf.iterrows():\n",
    "                ids = row.off_df.offtarget.values\n",
    "                exp_levels = exp_df['SUM_FPKM'][exp_df['transcript_id'].isin(ids)]\n",
    "                row.off_df.loc[:, 'sumexp'] = exp_levels\n",
    "                exp_values.append(sum(exp_levels))\n",
    "            self.probedf['sumexp'] = exp_values\n",
    "    def probeCountPerOfftarget(self, df, colname='offtarget'):\n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        offtarget_names = []\n",
    "        for ix, row in df.iterrows():\n",
    "            new_items = list(row.off_df[colname].values)\n",
    "            offtarget_names += new_items\n",
    "        c = Counter(offtarget_names)\n",
    "        return c\n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from __future__ import division, print_function\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "import pyensembl\n",
    "# from interval import interval\n",
    "\n",
    "import Bio.SeqIO\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "\n",
    "class GencodeGTF():\n",
    "    def __init__(self, gtfpath, fastapth):\n",
    "        self.df = pyensembl.GTF(gtfpath).dataframe()\n",
    "#         self._refseq = self._load_reference_fasta(fastapth)\n",
    "        self._parse_gene_symbol_from_transcript_name()\n",
    "#         self._trim_version_numbers()\n",
    "    def _load_reference_fasta(self, fastapth):\n",
    "        return Fasta(fastapth)\n",
    "    def _parse_gene_symbol_from_transcript_name(self):\n",
    "        symbol_list = []\n",
    "        for name in self.df['transcript_name'].values:\n",
    "            try:\n",
    "                gene_symbol = name.split('-')\n",
    "                if len(gene_symbol)==2:\n",
    "                    gene_symbol = gene_symbol[0]\n",
    "            except:\n",
    "                gene_symbol = np.nan\n",
    "            symbol_list.append(gene_symbol)\n",
    "        self.df['symbol'] = symbol_list\n",
    "    def _trim_version_numbers(self, trim_list = ['gene_id',\n",
    "                                                 'transcript_id']):\n",
    "        new_columns = [[] for i in range(len(trim_list))]\n",
    "        for idx, row in self.df.iterrows():\n",
    "            for list_idx, column in enumerate(trim_list):\n",
    "                try:\n",
    "                    name = row[column].split('.')\n",
    "                    if len(name)==2:\n",
    "                        name = name[0]\n",
    "                except:\n",
    "                    name = np.nan\n",
    "                new_columns[list_idx].append(name)\n",
    "        for idx, column in enumerate(trim_list):\n",
    "            self.df[column] = new_columns[idx]\n",
    "    def ref_lookup(self, seqname, start, end, strand='+'):\n",
    "        seq = self._refseq[seqname][start:end]\n",
    "        if strand=='-':\n",
    "            seq = seq.reverse.complement\n",
    "        return seq.seq\n",
    "    def find_isoform_ids(self, tid):\n",
    "        genename = self.df[self.df['transcript_id']==tid]['gene_id'].iloc[0]\n",
    "        x = self.df[self.df['gene_id'] == genename]['transcript_id'].unique()\n",
    "        x = [i for i in x if isinstance(i, str)]\n",
    "        return x\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "# gtf = GencodeGTF('/Users/rfor10/Downloads/gencode.v24.basic.annotation.gtf', \n",
    "#           '/Users/rfor10/Downloads/GRCh38.primary_assembly.genome.fa')\n",
    "# not_expressed = gtf.df[gtf.df.transcript_support_level!=1].transcript_id.unique()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Dataframe from /Users/rfor10/Desktop/Homo_sapiens.GRCh38.84.expanded.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rfor10/anaconda/lib/python3.5/site-packages/pyensembl/gtf.py:202: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  full_df = self._load_full_dataframe_cached()\n"
     ]
    }
   ],
   "source": [
    "gtf2 = GencodeGTF('/Users/rfor10/Desktop/Homo_sapiens.GRCh38.84.gtf', \n",
    "          '/Users/rfor10/Downloads/GRCh38.primary_assembly.genome.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>havana_transcript_version</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00002234944</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSE00003582793</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSE00002312635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00001948541</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSE00001671638</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSE00001758273</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSE00001799933</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSE00001746346</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSE00001863096</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00001890219</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSE00003507205</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSE00003477500</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSE00003565697</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSE00003475637</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSE00003502542</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSE00003553898</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSE00003621279</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSE00002030414</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>15005</td>\n",
       "      <td>15038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "      <td>ENSE00001935574</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>ENSE00001843071</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MIR6859, 1, 201]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003746039</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MIR6859, 1, 201]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[RP11, 34P13.3, 001]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569120</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>1529</td>\n",
       "      <td>1607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277107</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569121</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1529</td>\n",
       "      <td>1607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277107</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242209.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569122</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>1529</td>\n",
       "      <td>1607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277107</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003719399</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242209.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569123</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>5374</td>\n",
       "      <td>5469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274264</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569124</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5374</td>\n",
       "      <td>5469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274264</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242209.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569125</th>\n",
       "      <td>KI270724.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>5374</td>\n",
       "      <td>5469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274264</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003747774</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242209.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569126</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>32133</td>\n",
       "      <td>32228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276946</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569127</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>32133</td>\n",
       "      <td>32228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276946</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242006.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569128</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>32133</td>\n",
       "      <td>32228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276946</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003727624</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242006.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569129</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>36026</td>\n",
       "      <td>36107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278858</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569130</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>36026</td>\n",
       "      <td>36107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278858</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242006.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569131</th>\n",
       "      <td>KI270723.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>36026</td>\n",
       "      <td>36107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278858</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003715596</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC242006.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569132</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>1649</td>\n",
       "      <td>1731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277905</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569133</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1649</td>\n",
       "      <td>1731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277905</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FP325342.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569134</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>1649</td>\n",
       "      <td>1731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277905</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003749347</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FP325342.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569135</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3928</td>\n",
       "      <td>4011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000275028</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569136</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3928</td>\n",
       "      <td>4011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000275028</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FP325342.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569137</th>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>3928</td>\n",
       "      <td>4011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000275028</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003714321</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FP325342.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569138</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>144</td>\n",
       "      <td>268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278806</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569139</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>144</td>\n",
       "      <td>268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278806</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569140</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>144</td>\n",
       "      <td>268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278806</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003748419</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569141</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3497</td>\n",
       "      <td>3623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569142</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3497</td>\n",
       "      <td>3623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569143</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>3497</td>\n",
       "      <td>3623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274152</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003739940</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569144</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>9943</td>\n",
       "      <td>10067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276666</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569145</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>9943</td>\n",
       "      <td>10067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276666</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569146</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>9943</td>\n",
       "      <td>10067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000276666</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003754427</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569147</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21813</td>\n",
       "      <td>21944</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277917</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569148</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21813</td>\n",
       "      <td>21944</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277917</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569149</th>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21813</td>\n",
       "      <td>21944</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000277917</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003723556</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF065393.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2569150 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname   source     feature  start    end  score strand frame  \\\n",
       "0                 1   havana        gene  11869  14409    NaN      +     .   \n",
       "1                 1   havana  transcript  11869  14409    NaN      +     .   \n",
       "2                 1   havana        exon  11869  12227    NaN      +     .   \n",
       "3                 1   havana        exon  12613  12721    NaN      +     .   \n",
       "4                 1   havana        exon  13221  14409    NaN      +     .   \n",
       "5                 1   havana  transcript  12010  13670    NaN      +     .   \n",
       "6                 1   havana        exon  12010  12057    NaN      +     .   \n",
       "7                 1   havana        exon  12179  12227    NaN      +     .   \n",
       "8                 1   havana        exon  12613  12697    NaN      +     .   \n",
       "9                 1   havana        exon  12975  13052    NaN      +     .   \n",
       "10                1   havana        exon  13221  13374    NaN      +     .   \n",
       "11                1   havana        exon  13453  13670    NaN      +     .   \n",
       "12                1   havana        gene  14404  29570    NaN      -     .   \n",
       "13                1   havana  transcript  14404  29570    NaN      -     .   \n",
       "14                1   havana        exon  29534  29570    NaN      -     .   \n",
       "15                1   havana        exon  24738  24891    NaN      -     .   \n",
       "16                1   havana        exon  18268  18366    NaN      -     .   \n",
       "17                1   havana        exon  17915  18061    NaN      -     .   \n",
       "18                1   havana        exon  17606  17742    NaN      -     .   \n",
       "19                1   havana        exon  17233  17368    NaN      -     .   \n",
       "20                1   havana        exon  16858  17055    NaN      -     .   \n",
       "21                1   havana        exon  16607  16765    NaN      -     .   \n",
       "22                1   havana        exon  15796  15947    NaN      -     .   \n",
       "23                1   havana        exon  15005  15038    NaN      -     .   \n",
       "24                1   havana        exon  14404  14501    NaN      -     .   \n",
       "25                1  ensembl        gene  17369  17436    NaN      -     .   \n",
       "26                1  ensembl  transcript  17369  17436    NaN      -     .   \n",
       "27                1  ensembl        exon  17369  17436    NaN      -     .   \n",
       "28                1   havana        gene  29554  31109    NaN      +     .   \n",
       "29                1   havana  transcript  29554  31097    NaN      +     .   \n",
       "...             ...      ...         ...    ...    ...    ...    ...   ...   \n",
       "2569120  KI270724.1  ensembl        gene   1529   1607    NaN      -     .   \n",
       "2569121  KI270724.1  ensembl  transcript   1529   1607    NaN      -     .   \n",
       "2569122  KI270724.1  ensembl        exon   1529   1607    NaN      -     .   \n",
       "2569123  KI270724.1  ensembl        gene   5374   5469    NaN      +     .   \n",
       "2569124  KI270724.1  ensembl  transcript   5374   5469    NaN      +     .   \n",
       "2569125  KI270724.1  ensembl        exon   5374   5469    NaN      +     .   \n",
       "2569126  KI270723.1  ensembl        gene  32133  32228    NaN      -     .   \n",
       "2569127  KI270723.1  ensembl  transcript  32133  32228    NaN      -     .   \n",
       "2569128  KI270723.1  ensembl        exon  32133  32228    NaN      -     .   \n",
       "2569129  KI270723.1  ensembl        gene  36026  36107    NaN      +     .   \n",
       "2569130  KI270723.1  ensembl  transcript  36026  36107    NaN      +     .   \n",
       "2569131  KI270723.1  ensembl        exon  36026  36107    NaN      +     .   \n",
       "2569132  KI270707.1  ensembl        gene   1649   1731    NaN      -     .   \n",
       "2569133  KI270707.1  ensembl  transcript   1649   1731    NaN      -     .   \n",
       "2569134  KI270707.1  ensembl        exon   1649   1731    NaN      -     .   \n",
       "2569135  KI270707.1  ensembl        gene   3928   4011    NaN      -     .   \n",
       "2569136  KI270707.1  ensembl  transcript   3928   4011    NaN      -     .   \n",
       "2569137  KI270707.1  ensembl        exon   3928   4011    NaN      -     .   \n",
       "2569138  KI270752.1  ensembl        gene    144    268    NaN      +     .   \n",
       "2569139  KI270752.1  ensembl  transcript    144    268    NaN      +     .   \n",
       "2569140  KI270752.1  ensembl        exon    144    268    NaN      +     .   \n",
       "2569141  KI270752.1  ensembl        gene   3497   3623    NaN      -     .   \n",
       "2569142  KI270752.1  ensembl  transcript   3497   3623    NaN      -     .   \n",
       "2569143  KI270752.1  ensembl        exon   3497   3623    NaN      -     .   \n",
       "2569144  KI270752.1  ensembl        gene   9943  10067    NaN      -     .   \n",
       "2569145  KI270752.1  ensembl  transcript   9943  10067    NaN      -     .   \n",
       "2569146  KI270752.1  ensembl        exon   9943  10067    NaN      -     .   \n",
       "2569147  KI270752.1  ensembl        gene  21813  21944    NaN      +     .   \n",
       "2569148  KI270752.1  ensembl  transcript  21813  21944    NaN      +     .   \n",
       "2569149  KI270752.1  ensembl        exon  21813  21944    NaN      +     .   \n",
       "\n",
       "                 gene_id  gene_version          ...           \\\n",
       "0        ENSG00000223972             5          ...            \n",
       "1        ENSG00000223972             5          ...            \n",
       "2        ENSG00000223972             5          ...            \n",
       "3        ENSG00000223972             5          ...            \n",
       "4        ENSG00000223972             5          ...            \n",
       "5        ENSG00000223972             5          ...            \n",
       "6        ENSG00000223972             5          ...            \n",
       "7        ENSG00000223972             5          ...            \n",
       "8        ENSG00000223972             5          ...            \n",
       "9        ENSG00000223972             5          ...            \n",
       "10       ENSG00000223972             5          ...            \n",
       "11       ENSG00000223972             5          ...            \n",
       "12       ENSG00000227232             5          ...            \n",
       "13       ENSG00000227232             5          ...            \n",
       "14       ENSG00000227232             5          ...            \n",
       "15       ENSG00000227232             5          ...            \n",
       "16       ENSG00000227232             5          ...            \n",
       "17       ENSG00000227232             5          ...            \n",
       "18       ENSG00000227232             5          ...            \n",
       "19       ENSG00000227232             5          ...            \n",
       "20       ENSG00000227232             5          ...            \n",
       "21       ENSG00000227232             5          ...            \n",
       "22       ENSG00000227232             5          ...            \n",
       "23       ENSG00000227232             5          ...            \n",
       "24       ENSG00000227232             5          ...            \n",
       "25       ENSG00000278267             1          ...            \n",
       "26       ENSG00000278267             1          ...            \n",
       "27       ENSG00000278267             1          ...            \n",
       "28       ENSG00000243485             3          ...            \n",
       "29       ENSG00000243485             3          ...            \n",
       "...                  ...           ...          ...            \n",
       "2569120  ENSG00000277107             1          ...            \n",
       "2569121  ENSG00000277107             1          ...            \n",
       "2569122  ENSG00000277107             1          ...            \n",
       "2569123  ENSG00000274264             1          ...            \n",
       "2569124  ENSG00000274264             1          ...            \n",
       "2569125  ENSG00000274264             1          ...            \n",
       "2569126  ENSG00000276946             1          ...            \n",
       "2569127  ENSG00000276946             1          ...            \n",
       "2569128  ENSG00000276946             1          ...            \n",
       "2569129  ENSG00000278858             1          ...            \n",
       "2569130  ENSG00000278858             1          ...            \n",
       "2569131  ENSG00000278858             1          ...            \n",
       "2569132  ENSG00000277905             1          ...            \n",
       "2569133  ENSG00000277905             1          ...            \n",
       "2569134  ENSG00000277905             1          ...            \n",
       "2569135  ENSG00000275028             1          ...            \n",
       "2569136  ENSG00000275028             1          ...            \n",
       "2569137  ENSG00000275028             1          ...            \n",
       "2569138  ENSG00000278806             1          ...            \n",
       "2569139  ENSG00000278806             1          ...            \n",
       "2569140  ENSG00000278806             1          ...            \n",
       "2569141  ENSG00000274152             1          ...            \n",
       "2569142  ENSG00000274152             1          ...            \n",
       "2569143  ENSG00000274152             1          ...            \n",
       "2569144  ENSG00000276666             1          ...            \n",
       "2569145  ENSG00000276666             1          ...            \n",
       "2569146  ENSG00000276666             1          ...            \n",
       "2569147  ENSG00000277917             1          ...            \n",
       "2569148  ENSG00000277917             1          ...            \n",
       "2569149  ENSG00000277917             1          ...            \n",
       "\n",
       "        havana_transcript_version    tag transcript_support_level exon_number  \\\n",
       "0                             NaN    NaN                      NaN         NaN   \n",
       "1                             1.0  basic                      1.0         NaN   \n",
       "2                             1.0  basic                      1.0         1.0   \n",
       "3                             1.0  basic                      1.0         2.0   \n",
       "4                             1.0  basic                      1.0         3.0   \n",
       "5                             2.0  basic                      NaN         NaN   \n",
       "6                             2.0  basic                      NaN         1.0   \n",
       "7                             2.0  basic                      NaN         2.0   \n",
       "8                             2.0  basic                      NaN         3.0   \n",
       "9                             2.0  basic                      NaN         4.0   \n",
       "10                            2.0  basic                      NaN         5.0   \n",
       "11                            2.0  basic                      NaN         6.0   \n",
       "12                            NaN    NaN                      NaN         NaN   \n",
       "13                            1.0  basic                      NaN         NaN   \n",
       "14                            1.0  basic                      NaN         1.0   \n",
       "15                            1.0  basic                      NaN         2.0   \n",
       "16                            1.0  basic                      NaN         3.0   \n",
       "17                            1.0  basic                      NaN         4.0   \n",
       "18                            1.0  basic                      NaN         5.0   \n",
       "19                            1.0  basic                      NaN         6.0   \n",
       "20                            1.0  basic                      NaN         7.0   \n",
       "21                            1.0  basic                      NaN         8.0   \n",
       "22                            1.0  basic                      NaN         9.0   \n",
       "23                            1.0  basic                      NaN        10.0   \n",
       "24                            1.0  basic                      NaN        11.0   \n",
       "25                            NaN    NaN                      NaN         NaN   \n",
       "26                            NaN  basic                      NaN         NaN   \n",
       "27                            NaN  basic                      NaN         1.0   \n",
       "28                            NaN    NaN                      NaN         NaN   \n",
       "29                            1.0  basic                      5.0         NaN   \n",
       "...                           ...    ...                      ...         ...   \n",
       "2569120                       NaN    NaN                      NaN         NaN   \n",
       "2569121                       NaN  basic                      NaN         NaN   \n",
       "2569122                       NaN  basic                      NaN         1.0   \n",
       "2569123                       NaN    NaN                      NaN         NaN   \n",
       "2569124                       NaN  basic                      NaN         NaN   \n",
       "2569125                       NaN  basic                      NaN         1.0   \n",
       "2569126                       NaN    NaN                      NaN         NaN   \n",
       "2569127                       NaN  basic                      NaN         NaN   \n",
       "2569128                       NaN  basic                      NaN         1.0   \n",
       "2569129                       NaN    NaN                      NaN         NaN   \n",
       "2569130                       NaN  basic                      NaN         NaN   \n",
       "2569131                       NaN  basic                      NaN         1.0   \n",
       "2569132                       NaN    NaN                      NaN         NaN   \n",
       "2569133                       NaN  basic                      NaN         NaN   \n",
       "2569134                       NaN  basic                      NaN         1.0   \n",
       "2569135                       NaN    NaN                      NaN         NaN   \n",
       "2569136                       NaN  basic                      NaN         NaN   \n",
       "2569137                       NaN  basic                      NaN         1.0   \n",
       "2569138                       NaN    NaN                      NaN         NaN   \n",
       "2569139                       NaN  basic                      NaN         NaN   \n",
       "2569140                       NaN  basic                      NaN         1.0   \n",
       "2569141                       NaN    NaN                      NaN         NaN   \n",
       "2569142                       NaN  basic                      NaN         NaN   \n",
       "2569143                       NaN  basic                      NaN         1.0   \n",
       "2569144                       NaN    NaN                      NaN         NaN   \n",
       "2569145                       NaN  basic                      NaN         NaN   \n",
       "2569146                       NaN  basic                      NaN         1.0   \n",
       "2569147                       NaN    NaN                      NaN         NaN   \n",
       "2569148                       NaN  basic                      NaN         NaN   \n",
       "2569149                       NaN  basic                      NaN         1.0   \n",
       "\n",
       "                 exon_id exon_version  ccds_id protein_id protein_version  \\\n",
       "0                    NaN          NaN      NaN        NaN             NaN   \n",
       "1                    NaN          NaN      NaN        NaN             NaN   \n",
       "2        ENSE00002234944          1.0      NaN        NaN             NaN   \n",
       "3        ENSE00003582793          1.0      NaN        NaN             NaN   \n",
       "4        ENSE00002312635          1.0      NaN        NaN             NaN   \n",
       "5                    NaN          NaN      NaN        NaN             NaN   \n",
       "6        ENSE00001948541          1.0      NaN        NaN             NaN   \n",
       "7        ENSE00001671638          2.0      NaN        NaN             NaN   \n",
       "8        ENSE00001758273          2.0      NaN        NaN             NaN   \n",
       "9        ENSE00001799933          2.0      NaN        NaN             NaN   \n",
       "10       ENSE00001746346          2.0      NaN        NaN             NaN   \n",
       "11       ENSE00001863096          1.0      NaN        NaN             NaN   \n",
       "12                   NaN          NaN      NaN        NaN             NaN   \n",
       "13                   NaN          NaN      NaN        NaN             NaN   \n",
       "14       ENSE00001890219          1.0      NaN        NaN             NaN   \n",
       "15       ENSE00003507205          1.0      NaN        NaN             NaN   \n",
       "16       ENSE00003477500          1.0      NaN        NaN             NaN   \n",
       "17       ENSE00003565697          1.0      NaN        NaN             NaN   \n",
       "18       ENSE00003475637          1.0      NaN        NaN             NaN   \n",
       "19       ENSE00003502542          1.0      NaN        NaN             NaN   \n",
       "20       ENSE00003553898          1.0      NaN        NaN             NaN   \n",
       "21       ENSE00003621279          1.0      NaN        NaN             NaN   \n",
       "22       ENSE00002030414          1.0      NaN        NaN             NaN   \n",
       "23       ENSE00001935574          1.0      NaN        NaN             NaN   \n",
       "24       ENSE00001843071          1.0      NaN        NaN             NaN   \n",
       "25                   NaN          NaN      NaN        NaN             NaN   \n",
       "26                   NaN          NaN      NaN        NaN             NaN   \n",
       "27       ENSE00003746039          1.0      NaN        NaN             NaN   \n",
       "28                   NaN          NaN      NaN        NaN             NaN   \n",
       "29                   NaN          NaN      NaN        NaN             NaN   \n",
       "...                  ...          ...      ...        ...             ...   \n",
       "2569120              NaN          NaN      NaN        NaN             NaN   \n",
       "2569121              NaN          NaN      NaN        NaN             NaN   \n",
       "2569122  ENSE00003719399          1.0      NaN        NaN             NaN   \n",
       "2569123              NaN          NaN      NaN        NaN             NaN   \n",
       "2569124              NaN          NaN      NaN        NaN             NaN   \n",
       "2569125  ENSE00003747774          1.0      NaN        NaN             NaN   \n",
       "2569126              NaN          NaN      NaN        NaN             NaN   \n",
       "2569127              NaN          NaN      NaN        NaN             NaN   \n",
       "2569128  ENSE00003727624          1.0      NaN        NaN             NaN   \n",
       "2569129              NaN          NaN      NaN        NaN             NaN   \n",
       "2569130              NaN          NaN      NaN        NaN             NaN   \n",
       "2569131  ENSE00003715596          1.0      NaN        NaN             NaN   \n",
       "2569132              NaN          NaN      NaN        NaN             NaN   \n",
       "2569133              NaN          NaN      NaN        NaN             NaN   \n",
       "2569134  ENSE00003749347          1.0      NaN        NaN             NaN   \n",
       "2569135              NaN          NaN      NaN        NaN             NaN   \n",
       "2569136              NaN          NaN      NaN        NaN             NaN   \n",
       "2569137  ENSE00003714321          1.0      NaN        NaN             NaN   \n",
       "2569138              NaN          NaN      NaN        NaN             NaN   \n",
       "2569139              NaN          NaN      NaN        NaN             NaN   \n",
       "2569140  ENSE00003748419          1.0      NaN        NaN             NaN   \n",
       "2569141              NaN          NaN      NaN        NaN             NaN   \n",
       "2569142              NaN          NaN      NaN        NaN             NaN   \n",
       "2569143  ENSE00003739940          1.0      NaN        NaN             NaN   \n",
       "2569144              NaN          NaN      NaN        NaN             NaN   \n",
       "2569145              NaN          NaN      NaN        NaN             NaN   \n",
       "2569146  ENSE00003754427          1.0      NaN        NaN             NaN   \n",
       "2569147              NaN          NaN      NaN        NaN             NaN   \n",
       "2569148              NaN          NaN      NaN        NaN             NaN   \n",
       "2569149  ENSE00003723556          1.0      NaN        NaN             NaN   \n",
       "\n",
       "                       symbol  \n",
       "0                         NaN  \n",
       "1                     DDX11L1  \n",
       "2                     DDX11L1  \n",
       "3                     DDX11L1  \n",
       "4                     DDX11L1  \n",
       "5                     DDX11L1  \n",
       "6                     DDX11L1  \n",
       "7                     DDX11L1  \n",
       "8                     DDX11L1  \n",
       "9                     DDX11L1  \n",
       "10                    DDX11L1  \n",
       "11                    DDX11L1  \n",
       "12                        NaN  \n",
       "13                     WASH7P  \n",
       "14                     WASH7P  \n",
       "15                     WASH7P  \n",
       "16                     WASH7P  \n",
       "17                     WASH7P  \n",
       "18                     WASH7P  \n",
       "19                     WASH7P  \n",
       "20                     WASH7P  \n",
       "21                     WASH7P  \n",
       "22                     WASH7P  \n",
       "23                     WASH7P  \n",
       "24                     WASH7P  \n",
       "25                        NaN  \n",
       "26          [MIR6859, 1, 201]  \n",
       "27          [MIR6859, 1, 201]  \n",
       "28                        NaN  \n",
       "29       [RP11, 34P13.3, 001]  \n",
       "...                       ...  \n",
       "2569120                   NaN  \n",
       "2569121            AC242209.2  \n",
       "2569122            AC242209.2  \n",
       "2569123                   NaN  \n",
       "2569124            AC242209.1  \n",
       "2569125            AC242209.1  \n",
       "2569126                   NaN  \n",
       "2569127            AC242006.1  \n",
       "2569128            AC242006.1  \n",
       "2569129                   NaN  \n",
       "2569130            AC242006.2  \n",
       "2569131            AC242006.2  \n",
       "2569132                   NaN  \n",
       "2569133            FP325342.2  \n",
       "2569134            FP325342.2  \n",
       "2569135                   NaN  \n",
       "2569136            FP325342.1  \n",
       "2569137            FP325342.1  \n",
       "2569138                   NaN  \n",
       "2569139            AF065393.4  \n",
       "2569140            AF065393.4  \n",
       "2569141                   NaN  \n",
       "2569142            AF065393.1  \n",
       "2569143            AF065393.1  \n",
       "2569144                   NaN  \n",
       "2569145            AF065393.2  \n",
       "2569146            AF065393.2  \n",
       "2569147                   NaN  \n",
       "2569148            AF065393.3  \n",
       "2569149            AF065393.3  \n",
       "\n",
       "[2569150 rows x 31 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf2.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'float' object has no attribute 'split'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-50e503b048ec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnewdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'transcript'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnewdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'transcript_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-16-50e503b048ec>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnewdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'transcript'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnewdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'transcript_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'float' object has no attribute 'split'"
     ]
    }
   ],
   "source": [
    "newdf['transcript'] = [t.split('.')[0] for t in newdf['transcript_id'].values if isinstance]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "11000\n",
      "44000\n",
      "71000\n",
      "82000\n",
      "90000\n",
      "112000\n",
      "115000\n",
      "118000\n",
      "168000\n",
      "170000\n",
      "206000\n",
      "214000\n",
      "215000\n",
      "257000\n",
      "260000\n",
      "262000\n",
      "270000\n",
      "290000\n",
      "304000\n",
      "319000\n",
      "326000\n",
      "330000\n",
      "342000\n",
      "384000\n",
      "393000\n",
      "400000\n",
      "416000\n",
      "425000\n",
      "436000\n",
      "440000\n",
      "455000\n",
      "473000\n",
      "491000\n",
      "495000\n",
      "498000\n",
      "504000\n",
      "515000\n",
      "516000\n",
      "543000\n",
      "557000\n",
      "563000\n",
      "575000\n",
      "577000\n",
      "590000\n",
      "603000\n",
      "607000\n",
      "614000\n",
      "617000\n",
      "669000\n",
      "673000\n",
      "699000\n",
      "701000\n",
      "725000\n",
      "726000\n",
      "733000\n",
      "738000\n",
      "752000\n",
      "755000\n",
      "770000\n",
      "794000\n",
      "802000\n",
      "819000\n",
      "839000\n",
      "844000\n",
      "867000\n",
      "884000\n",
      "890000\n",
      "891000\n",
      "907000\n",
      "929000\n",
      "936000\n",
      "944000\n",
      "949000\n",
      "974000\n",
      "982000\n",
      "989000\n",
      "991000\n",
      "1021000\n",
      "1027000\n",
      "1033000\n",
      "1046000\n",
      "1049000\n",
      "1057000\n",
      "1080000\n",
      "1085000\n",
      "1089000\n",
      "1094000\n",
      "1097000\n",
      "1107000\n",
      "1109000\n",
      "1119000\n",
      "1138000\n",
      "1143000\n",
      "1150000\n",
      "1154000\n",
      "1164000\n",
      "1165000\n",
      "1171000\n",
      "1178000\n",
      "1185000\n",
      "1211000\n",
      "1219000\n",
      "1226000\n",
      "1248000\n",
      "1269000\n",
      "1277000\n",
      "1285000\n",
      "1290000\n",
      "1292000\n",
      "1312000\n",
      "1313000\n",
      "1320000\n",
      "1330000\n",
      "1358000\n",
      "1360000\n",
      "1368000\n",
      "1377000\n",
      "1384000\n",
      "1404000\n",
      "1414000\n",
      "1423000\n",
      "1429000\n",
      "1430000\n",
      "1456000\n",
      "1522000\n",
      "1523000\n",
      "1539000\n",
      "1549000\n",
      "1551000\n",
      "1567000\n",
      "1583000\n",
      "1621000\n",
      "1636000\n",
      "1640000\n",
      "1645000\n",
      "1648000\n",
      "1668000\n",
      "1695000\n",
      "1710000\n",
      "1719000\n",
      "1720000\n",
      "1722000\n",
      "1726000\n",
      "1736000\n",
      "1743000\n",
      "1852000\n",
      "1875000\n",
      "1879000\n",
      "1881000\n",
      "1885000\n",
      "1889000\n",
      "1906000\n",
      "1907000\n",
      "1922000\n",
      "1924000\n",
      "1931000\n",
      "1936000\n",
      "1952000\n",
      "1973000\n",
      "1990000\n",
      "2017000\n",
      "2039000\n",
      "2042000\n",
      "2048000\n",
      "2082000\n",
      "2085000\n",
      "2087000\n",
      "2103000\n",
      "2118000\n",
      "2131000\n",
      "2139000\n",
      "2156000\n",
      "2159000\n",
      "2165000\n",
      "2172000\n",
      "2197000\n",
      "2199000\n",
      "2204000\n",
      "2213000\n",
      "2216000\n",
      "2228000\n",
      "2243000\n",
      "2246000\n",
      "2247000\n",
      "2259000\n",
      "2282000\n",
      "2322000\n",
      "2326000\n",
      "2414000\n",
      "2417000\n",
      "2425000\n",
      "2454000\n",
      "2458000\n",
      "2472000\n",
      "2476000\n",
      "2485000\n",
      "2507000\n",
      "2523000\n",
      "2532000\n",
      "2537000\n",
      "2548000\n",
      "2568000\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "gene_name_to_transcripts = defaultdict(list)\n",
    "transcript_to_gene = defaultdict(list)\n",
    "newdf = gtf2.df[['gene_id', 'transcript_id', 'transcript_name', 'gene_name']].drop_duplicates(subset='transcript_id')\n",
    "for idx, row in newdf.iterrows():\n",
    "    if idx % 1000 == 0:\n",
    "        print(idx)\n",
    "    gene_id, transcript_id, tname, gname = row\n",
    "    if isinstance(tname, float):\n",
    "        continue\n",
    "    gene_id = gene_id.split('.')[0]\n",
    "    gene_symbol = gname\n",
    "    transcript_id = transcript_id.split('.')[0]\n",
    "    if transcript_id not in gene_name_to_transcripts[gene_id]:\n",
    "        gene_name_to_transcripts[gene_id].append(transcript_id)\n",
    "        gene_name_to_transcripts[gene_symbol].append(transcript_id)\n",
    "    \n",
    "    transcript_to_gene[transcript_id] =((gene_symbol, gene_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "pickle.dump(gene_name_to_transcripts, open('/data/wollman/hypefish/datafiles/gene2transcripts.pck', 'wb'))\n",
    "pickle.dump(transcript_to_gene, open('/data/wollman/hypefish/datafiles/transcripts2genes.pck', 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SeqRecord(seq=Seq('ATGAGTGCGATTAAGCCAGACATGAAGATCAACCTCCGTATGGAAGGCAACGTA...TAA', SingleLetterAlphabet()), id='tdEOS', name='tdEOS', description='tdEOS', dbxrefs=[])]\n",
      "\n",
      "\t***\tOligoArray 2.1.3\t***\n",
      "\n",
      "OligoArray 2.1.3 will start to process sequences from the file /data/a28cedaf-ce2f-4acb-a05e-d40c3e2a1a8c/temp.fa using the following parameters :\n",
      "Blast database: '/data/a28cedaf-ce2f-4acb-a05e-d40c3e2a1a8c/db.fa'\n",
      "Oligo data will be saved in: '/data/a28cedaf-ce2f-4acb-a05e-d40c3e2a1a8c/oligos.txt'\n",
      "Sequence without oligo will be saved in: 'rejected.fas'\n",
      "The log file will be: '/data/a28cedaf-ce2f-4acb-a05e-d40c3e2a1a8c/oligos.log'\n",
      "Maximum number of oligo to design per input sequence: '196'\n",
      "Size range: '30' to '30'\n",
      "Maximum distance between the 5' end of the oligo and the 3' end of the input sequence: '10000'\n",
      "Minimum distance between the 5' ends of two adjacent oligos: '30'\n",
      "Tm range: '70' to '100'\n",
      "GC range: '20.0' to '80.0'\n",
      "Threshold to reject secondary structures: '76.0'\n",
      "Threshold to start to consider cross-hybridizations: '72'\n",
      "Sequence to avoid in the oligo: 'GGGGGG;CCCCCC;TTTTTT;AAAAAA'\n",
      "Number of sequence to run in parallel: '16'\n",
      "\n",
      "Can OligoArray read/write specified files?  YES\n",
      "Data initialization: DONE\n",
      "Is /data/a28cedaf-ce2f-4acb-a05e-d40c3e2a1a8c/db.fa a valid Blast database?  YES\n",
      "Is OligoArrayAux installed?  YES\n",
      "\n",
      "\n",
      "Start Blast parameters initialization (It may take a while depending the value entered for the -D option)\n",
      "Blast parameters initialized\n",
      "\n",
      "No more sequence to dispatch\n",
      "Start tdEOS\n",
      "OligoArray has successfully processed all sequences\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    d = DesignTargets()\n",
    "    # ENST00000356805.8 is the tid of SPTBN1 gene\n",
    "    d.designProbes(transcript_id=['tdEOS'])\n",
    "    on = d._parseOligoArrayOutput()\n",
    "#     targets = ScreenTargets(on)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>pos</th>\n",
       "      <th>len</th>\n",
       "      <th>dG</th>\n",
       "      <th>dH</th>\n",
       "      <th>dS</th>\n",
       "      <th>Tm</th>\n",
       "      <th>seq</th>\n",
       "      <th>off_df</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1486</td>\n",
       "      <td>30</td>\n",
       "      <td>-179.59</td>\n",
       "      <td>-226.39</td>\n",
       "      <td>-615.89</td>\n",
       "      <td>77.25</td>\n",
       "      <td>AGAAAGGTAGGATCCACCGGATCTAGATAA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1398</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.74</td>\n",
       "      <td>-239.59</td>\n",
       "      <td>-642.8</td>\n",
       "      <td>82.86</td>\n",
       "      <td>ACGATCCGGACTCAGATCTCGAGCTGATCC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1368</td>\n",
       "      <td>30</td>\n",
       "      <td>-188.34</td>\n",
       "      <td>-236.70</td>\n",
       "      <td>-636.19</td>\n",
       "      <td>82.03</td>\n",
       "      <td>TGCTCATTCTGGATTGCCTGACAATGCCAG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1338</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.08</td>\n",
       "      <td>-235.7</td>\n",
       "      <td>-639.69</td>\n",
       "      <td>78.68</td>\n",
       "      <td>CAACAAGGTTAAGCTGTATGAGCATGCTGT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1307</td>\n",
       "      <td>30</td>\n",
       "      <td>-183.19</td>\n",
       "      <td>-231.5</td>\n",
       "      <td>-635.59</td>\n",
       "      <td>74.54</td>\n",
       "      <td>GCATTGAGATTTTAAGCCATGACAAAGATT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1277</td>\n",
       "      <td>30</td>\n",
       "      <td>-184.49</td>\n",
       "      <td>-231.8</td>\n",
       "      <td>-622.39</td>\n",
       "      <td>82.03</td>\n",
       "      <td>AGTTACCAGGCTACCACTTTGTGGACCACT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1247</td>\n",
       "      <td>30</td>\n",
       "      <td>-182.99</td>\n",
       "      <td>-230.8</td>\n",
       "      <td>-628.99</td>\n",
       "      <td>76.96</td>\n",
       "      <td>CTACTTACAAAGCTAAGGAGAAGGGTGTCA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1217</td>\n",
       "      <td>30</td>\n",
       "      <td>-186.17</td>\n",
       "      <td>-234.6</td>\n",
       "      <td>-637.19</td>\n",
       "      <td>78.35</td>\n",
       "      <td>GAAATGCCCATTACCGATGTGACTTCAGAA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1187</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.23</td>\n",
       "      <td>-239.80</td>\n",
       "      <td>-652.19</td>\n",
       "      <td>78.25</td>\n",
       "      <td>GTGATATTCGCATGGCTTTGTTGCTTGAAG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1157</td>\n",
       "      <td>30</td>\n",
       "      <td>-189.40</td>\n",
       "      <td>-238.40</td>\n",
       "      <td>-644.69</td>\n",
       "      <td>80.08</td>\n",
       "      <td>AAATGTATGTGCGTGATGGAGTGCTGACTG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1100</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.94</td>\n",
       "      <td>-236.3</td>\n",
       "      <td>-636.3</td>\n",
       "      <td>81.38</td>\n",
       "      <td>CCGCCAATGGTCCAGTTATGCAGAAGAAGA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1070</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.12</td>\n",
       "      <td>-234.10</td>\n",
       "      <td>-644.4</td>\n",
       "      <td>73.86</td>\n",
       "      <td>ATAAAGTTCGATTTCACGGTGTAAACTTTC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1040</td>\n",
       "      <td>30</td>\n",
       "      <td>-178.76</td>\n",
       "      <td>-225.69</td>\n",
       "      <td>-617.49</td>\n",
       "      <td>75.30</td>\n",
       "      <td>ACATAACAATGGAAGGGGACACTTTCTATA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>1010</td>\n",
       "      <td>30</td>\n",
       "      <td>-192.31</td>\n",
       "      <td>-241.30</td>\n",
       "      <td>-644.5</td>\n",
       "      <td>84.48</td>\n",
       "      <td>AAGACGGGGGCATTTGCATTGCCAGAAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>966</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.49</td>\n",
       "      <td>-239.59</td>\n",
       "      <td>-646.09</td>\n",
       "      <td>81.12</td>\n",
       "      <td>GTCGTTTCCTAAGGGGTATTCGTGGGAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>936</td>\n",
       "      <td>30</td>\n",
       "      <td>-181.06</td>\n",
       "      <td>-228.60</td>\n",
       "      <td>-625.39</td>\n",
       "      <td>75.53</td>\n",
       "      <td>TCCAGACCACATACAAGACTATTTTAAGCA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>902</td>\n",
       "      <td>30</td>\n",
       "      <td>-192.68</td>\n",
       "      <td>-242.2</td>\n",
       "      <td>-651.49</td>\n",
       "      <td>82.13</td>\n",
       "      <td>CATTCCATTACGGCAACAGGGTATTCGCCG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>872</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.88</td>\n",
       "      <td>-233.89</td>\n",
       "      <td>-631.79</td>\n",
       "      <td>80.16</td>\n",
       "      <td>TGCCTTTTGCCTTTGATATCCTGACCACTG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>842</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.44</td>\n",
       "      <td>-235.5</td>\n",
       "      <td>-632.29</td>\n",
       "      <td>82.31</td>\n",
       "      <td>TGGATCTTGAAGTCAAAGAGGGCGGACCTC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>812</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.86</td>\n",
       "      <td>-233.90</td>\n",
       "      <td>-631.99</td>\n",
       "      <td>80.06</td>\n",
       "      <td>CAGGCAAGCCTTTTGAGGGAAAACAGAGTA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>782</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.30</td>\n",
       "      <td>-239.09</td>\n",
       "      <td>-642.09</td>\n",
       "      <td>82.49</td>\n",
       "      <td>GGCACCACTTTGTGATCGACGGAGATGGTA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>752</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.82</td>\n",
       "      <td>-240.00</td>\n",
       "      <td>-646.99</td>\n",
       "      <td>81.24</td>\n",
       "      <td>TCAACCTCCGTATGGAAGGCAACGTAAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>722</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.69</td>\n",
       "      <td>-236.4</td>\n",
       "      <td>-640.8</td>\n",
       "      <td>79.15</td>\n",
       "      <td>CTGAGAGTGCGATTAAGCCAGACATGAAGA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>692</td>\n",
       "      <td>30</td>\n",
       "      <td>-184.28</td>\n",
       "      <td>-231.90</td>\n",
       "      <td>-626.5</td>\n",
       "      <td>79.97</td>\n",
       "      <td>ACGGTACTGGTTCTACTGGTTCTGGTTCTT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>662</td>\n",
       "      <td>30</td>\n",
       "      <td>-194.97</td>\n",
       "      <td>-244.7</td>\n",
       "      <td>-654.29</td>\n",
       "      <td>84.33</td>\n",
       "      <td>CTGACAATGCCAGACGATCCGGTAGCGGTC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>632</td>\n",
       "      <td>30</td>\n",
       "      <td>-188.96</td>\n",
       "      <td>-237.7</td>\n",
       "      <td>-641.19</td>\n",
       "      <td>80.88</td>\n",
       "      <td>ATGAGCATGCTGTTGCTCATTCTGGATTGC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>602</td>\n",
       "      <td>30</td>\n",
       "      <td>-181.61</td>\n",
       "      <td>-229.40</td>\n",
       "      <td>-628.79</td>\n",
       "      <td>74.94</td>\n",
       "      <td>ATGACAAAGATTACAACAAGGTTAAGCTGT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>572</td>\n",
       "      <td>30</td>\n",
       "      <td>-186.10</td>\n",
       "      <td>-234.4</td>\n",
       "      <td>-635.49</td>\n",
       "      <td>78.95</td>\n",
       "      <td>TTGTGGACCACTGCATTGAGATTTTAAGCC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>542</td>\n",
       "      <td>30</td>\n",
       "      <td>-183.74</td>\n",
       "      <td>-230.90</td>\n",
       "      <td>-620.49</td>\n",
       "      <td>81.69</td>\n",
       "      <td>AGAAGGGTGTCAAGTTACCAGGCTACCACT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>512</td>\n",
       "      <td>30</td>\n",
       "      <td>-184.60</td>\n",
       "      <td>-233.30</td>\n",
       "      <td>-640.69</td>\n",
       "      <td>74.58</td>\n",
       "      <td>GTGACTTCAGAACTACTTACAAAGCTAAGG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>482</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.41</td>\n",
       "      <td>-233.29</td>\n",
       "      <td>-629.99</td>\n",
       "      <td>80.22</td>\n",
       "      <td>TGTTGCTTGAAGGAAATGCCCATTACCGAT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>452</td>\n",
       "      <td>30</td>\n",
       "      <td>-189.60</td>\n",
       "      <td>-238.4</td>\n",
       "      <td>-641.99</td>\n",
       "      <td>81.50</td>\n",
       "      <td>GAGTGCTGACTGGTGATATTCGCATGGCTT</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>422</td>\n",
       "      <td>30</td>\n",
       "      <td>-188.87</td>\n",
       "      <td>-238.0</td>\n",
       "      <td>-646.39</td>\n",
       "      <td>78.60</td>\n",
       "      <td>CCTCCACTGAGAAAATGTATGTGCGTGATG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>377</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.94</td>\n",
       "      <td>-236.3</td>\n",
       "      <td>-636.3</td>\n",
       "      <td>81.38</td>\n",
       "      <td>CCGCCAATGGTCCAGTTATGCAGAAGAAGA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>347</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.12</td>\n",
       "      <td>-234.10</td>\n",
       "      <td>-644.4</td>\n",
       "      <td>73.86</td>\n",
       "      <td>ATAAAGTTCGATTTCACGGTGTAAACTTTC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>317</td>\n",
       "      <td>30</td>\n",
       "      <td>-178.76</td>\n",
       "      <td>-225.69</td>\n",
       "      <td>-617.49</td>\n",
       "      <td>75.30</td>\n",
       "      <td>ACATAACAATGGAAGGGGACACTTTCTATA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>287</td>\n",
       "      <td>30</td>\n",
       "      <td>-192.31</td>\n",
       "      <td>-241.30</td>\n",
       "      <td>-644.5</td>\n",
       "      <td>84.48</td>\n",
       "      <td>AAGACGGGGGCATTTGCATTGCCAGAAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>243</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.49</td>\n",
       "      <td>-239.59</td>\n",
       "      <td>-646.09</td>\n",
       "      <td>81.12</td>\n",
       "      <td>GTCGTTTCCTAAGGGGTATTCGTGGGAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>213</td>\n",
       "      <td>30</td>\n",
       "      <td>-181.06</td>\n",
       "      <td>-228.60</td>\n",
       "      <td>-625.39</td>\n",
       "      <td>75.53</td>\n",
       "      <td>TCCAGACCACATACAAGACTATTTTAAGCA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>179</td>\n",
       "      <td>30</td>\n",
       "      <td>-192.68</td>\n",
       "      <td>-242.2</td>\n",
       "      <td>-651.49</td>\n",
       "      <td>82.13</td>\n",
       "      <td>CATTCCATTACGGCAACAGGGTATTCGCCG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>149</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.88</td>\n",
       "      <td>-233.89</td>\n",
       "      <td>-631.79</td>\n",
       "      <td>80.16</td>\n",
       "      <td>TGCCTTTTGCCTTTGATATCCTGACCACTG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>119</td>\n",
       "      <td>30</td>\n",
       "      <td>-187.44</td>\n",
       "      <td>-235.5</td>\n",
       "      <td>-632.29</td>\n",
       "      <td>82.31</td>\n",
       "      <td>TGGATCTTGAAGTCAAAGAGGGCGGACCTC</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>89</td>\n",
       "      <td>30</td>\n",
       "      <td>-185.86</td>\n",
       "      <td>-233.90</td>\n",
       "      <td>-631.99</td>\n",
       "      <td>80.06</td>\n",
       "      <td>CAGGCAAGCCTTTTGAGGGAAAACAGAGTA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.30</td>\n",
       "      <td>-239.09</td>\n",
       "      <td>-642.09</td>\n",
       "      <td>82.49</td>\n",
       "      <td>GGCACCACTTTGTGATCGACGGAGATGGTA</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>tdEOS</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>-190.82</td>\n",
       "      <td>-240.00</td>\n",
       "      <td>-646.99</td>\n",
       "      <td>81.24</td>\n",
       "      <td>TCAACCTCCGTATGGAAGGCAACGTAAACG</td>\n",
       "      <td>Empty DataFrame\n",
       "Columns: [target, pos, offtarg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target   pos len       dG       dH       dS     Tm  \\\n",
       "0   tdEOS  1486  30  -179.59  -226.39  -615.89  77.25   \n",
       "1   tdEOS  1398  30  -190.74  -239.59   -642.8  82.86   \n",
       "2   tdEOS  1368  30  -188.34  -236.70  -636.19  82.03   \n",
       "3   tdEOS  1338  30  -187.08   -235.7  -639.69  78.68   \n",
       "4   tdEOS  1307  30  -183.19   -231.5  -635.59  74.54   \n",
       "5   tdEOS  1277  30  -184.49   -231.8  -622.39  82.03   \n",
       "6   tdEOS  1247  30  -182.99   -230.8  -628.99  76.96   \n",
       "7   tdEOS  1217  30  -186.17   -234.6  -637.19  78.35   \n",
       "8   tdEOS  1187  30  -190.23  -239.80  -652.19  78.25   \n",
       "9   tdEOS  1157  30  -189.40  -238.40  -644.69  80.08   \n",
       "10  tdEOS  1100  30  -187.94   -236.3   -636.3  81.38   \n",
       "11  tdEOS  1070  30  -185.12  -234.10   -644.4  73.86   \n",
       "12  tdEOS  1040  30  -178.76  -225.69  -617.49  75.30   \n",
       "13  tdEOS  1010  30  -192.31  -241.30   -644.5  84.48   \n",
       "14  tdEOS   966  30  -190.49  -239.59  -646.09  81.12   \n",
       "15  tdEOS   936  30  -181.06  -228.60  -625.39  75.53   \n",
       "16  tdEOS   902  30  -192.68   -242.2  -651.49  82.13   \n",
       "17  tdEOS   872  30  -185.88  -233.89  -631.79  80.16   \n",
       "18  tdEOS   842  30  -187.44   -235.5  -632.29  82.31   \n",
       "19  tdEOS   812  30  -185.86  -233.90  -631.99  80.06   \n",
       "20  tdEOS   782  30  -190.30  -239.09  -642.09  82.49   \n",
       "21  tdEOS   752  30  -190.82  -240.00  -646.99  81.24   \n",
       "22  tdEOS   722  30  -187.69   -236.4   -640.8  79.15   \n",
       "23  tdEOS   692  30  -184.28  -231.90   -626.5  79.97   \n",
       "24  tdEOS   662  30  -194.97   -244.7  -654.29  84.33   \n",
       "25  tdEOS   632  30  -188.96   -237.7  -641.19  80.88   \n",
       "26  tdEOS   602  30  -181.61  -229.40  -628.79  74.94   \n",
       "27  tdEOS   572  30  -186.10   -234.4  -635.49  78.95   \n",
       "28  tdEOS   542  30  -183.74  -230.90  -620.49  81.69   \n",
       "29  tdEOS   512  30  -184.60  -233.30  -640.69  74.58   \n",
       "30  tdEOS   482  30  -185.41  -233.29  -629.99  80.22   \n",
       "31  tdEOS   452  30  -189.60   -238.4  -641.99  81.50   \n",
       "32  tdEOS   422  30  -188.87   -238.0  -646.39  78.60   \n",
       "33  tdEOS   377  30  -187.94   -236.3   -636.3  81.38   \n",
       "34  tdEOS   347  30  -185.12  -234.10   -644.4  73.86   \n",
       "35  tdEOS   317  30  -178.76  -225.69  -617.49  75.30   \n",
       "36  tdEOS   287  30  -192.31  -241.30   -644.5  84.48   \n",
       "37  tdEOS   243  30  -190.49  -239.59  -646.09  81.12   \n",
       "38  tdEOS   213  30  -181.06  -228.60  -625.39  75.53   \n",
       "39  tdEOS   179  30  -192.68   -242.2  -651.49  82.13   \n",
       "40  tdEOS   149  30  -185.88  -233.89  -631.79  80.16   \n",
       "41  tdEOS   119  30  -187.44   -235.5  -632.29  82.31   \n",
       "42  tdEOS    89  30  -185.86  -233.90  -631.99  80.06   \n",
       "43  tdEOS    59  30  -190.30  -239.09  -642.09  82.49   \n",
       "44  tdEOS    29  30  -190.82  -240.00  -646.99  81.24   \n",
       "\n",
       "                               seq  \\\n",
       "0   AGAAAGGTAGGATCCACCGGATCTAGATAA   \n",
       "1   ACGATCCGGACTCAGATCTCGAGCTGATCC   \n",
       "2   TGCTCATTCTGGATTGCCTGACAATGCCAG   \n",
       "3   CAACAAGGTTAAGCTGTATGAGCATGCTGT   \n",
       "4   GCATTGAGATTTTAAGCCATGACAAAGATT   \n",
       "5   AGTTACCAGGCTACCACTTTGTGGACCACT   \n",
       "6   CTACTTACAAAGCTAAGGAGAAGGGTGTCA   \n",
       "7   GAAATGCCCATTACCGATGTGACTTCAGAA   \n",
       "8   GTGATATTCGCATGGCTTTGTTGCTTGAAG   \n",
       "9   AAATGTATGTGCGTGATGGAGTGCTGACTG   \n",
       "10  CCGCCAATGGTCCAGTTATGCAGAAGAAGA   \n",
       "11  ATAAAGTTCGATTTCACGGTGTAAACTTTC   \n",
       "12  ACATAACAATGGAAGGGGACACTTTCTATA   \n",
       "13  AAGACGGGGGCATTTGCATTGCCAGAAACG   \n",
       "14  GTCGTTTCCTAAGGGGTATTCGTGGGAACG   \n",
       "15  TCCAGACCACATACAAGACTATTTTAAGCA   \n",
       "16  CATTCCATTACGGCAACAGGGTATTCGCCG   \n",
       "17  TGCCTTTTGCCTTTGATATCCTGACCACTG   \n",
       "18  TGGATCTTGAAGTCAAAGAGGGCGGACCTC   \n",
       "19  CAGGCAAGCCTTTTGAGGGAAAACAGAGTA   \n",
       "20  GGCACCACTTTGTGATCGACGGAGATGGTA   \n",
       "21  TCAACCTCCGTATGGAAGGCAACGTAAACG   \n",
       "22  CTGAGAGTGCGATTAAGCCAGACATGAAGA   \n",
       "23  ACGGTACTGGTTCTACTGGTTCTGGTTCTT   \n",
       "24  CTGACAATGCCAGACGATCCGGTAGCGGTC   \n",
       "25  ATGAGCATGCTGTTGCTCATTCTGGATTGC   \n",
       "26  ATGACAAAGATTACAACAAGGTTAAGCTGT   \n",
       "27  TTGTGGACCACTGCATTGAGATTTTAAGCC   \n",
       "28  AGAAGGGTGTCAAGTTACCAGGCTACCACT   \n",
       "29  GTGACTTCAGAACTACTTACAAAGCTAAGG   \n",
       "30  TGTTGCTTGAAGGAAATGCCCATTACCGAT   \n",
       "31  GAGTGCTGACTGGTGATATTCGCATGGCTT   \n",
       "32  CCTCCACTGAGAAAATGTATGTGCGTGATG   \n",
       "33  CCGCCAATGGTCCAGTTATGCAGAAGAAGA   \n",
       "34  ATAAAGTTCGATTTCACGGTGTAAACTTTC   \n",
       "35  ACATAACAATGGAAGGGGACACTTTCTATA   \n",
       "36  AAGACGGGGGCATTTGCATTGCCAGAAACG   \n",
       "37  GTCGTTTCCTAAGGGGTATTCGTGGGAACG   \n",
       "38  TCCAGACCACATACAAGACTATTTTAAGCA   \n",
       "39  CATTCCATTACGGCAACAGGGTATTCGCCG   \n",
       "40  TGCCTTTTGCCTTTGATATCCTGACCACTG   \n",
       "41  TGGATCTTGAAGTCAAAGAGGGCGGACCTC   \n",
       "42  CAGGCAAGCCTTTTGAGGGAAAACAGAGTA   \n",
       "43  GGCACCACTTTGTGATCGACGGAGATGGTA   \n",
       "44  TCAACCTCCGTATGGAAGGCAACGTAAACG   \n",
       "\n",
       "                                               off_df  \n",
       "0   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "1   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "2   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "3   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "4   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "5   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "6   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "7   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "8   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "9   Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "10  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "11  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "12  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "13  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "14  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "15  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "16  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "17  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "18  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "19  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "20  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "21  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "22  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "23  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "24  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "25  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "26  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "27  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "28  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "29  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "30  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "31  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "32  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "33  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "34  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "35  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "36  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "37  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "38  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "39  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "40  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "41  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "42  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "43  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  \n",
       "44  Empty DataFrame\n",
       "Columns: [target, pos, offtarg...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "on"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "alt_map = {'ins':'0'}\n",
    "complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'} \n",
    "\n",
    "def reverse_complement(seq):    \n",
    "    for k,v in alt_map.iteritems():\n",
    "        seq = seq.replace(k,v)\n",
    "    bases = list(seq) \n",
    "    bases = reversed([complement.get(base,base) for base in bases])\n",
    "    bases = ''.join(bases)\n",
    "    for k,v in alt_map.iteritems():\n",
    "        bases = bases.replace(v,k)\n",
    "    return bases\n",
    "\n",
    "# print([reverse_complement(i) for i in on.seq.values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import cPickle as pickle\n",
    "exp_df = pickle.load(open('/data/wollman/hypefish/ipython/research-python/imr90_combined_df.pck', 'r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def aggOfftargetExpression(off_df, on_df, expression_df, linker='id',\n",
    "                           expression_column='SUM_FPKM', aggfunc = sum, arg = None):\n",
    "    \"\"\"\n",
    "    Given a groupby objected formed on the linker column name and \n",
    "    an expression_df with the column expression_column. Produce \n",
    "    an output where group id's looked up in expression_df and \n",
    "    values are aggregated by aggfunc.\n",
    "    \"\"\"\n",
    "    if arg is None:\n",
    "        arg = {'tm_thresh': 70, \n",
    "               }\n",
    "    on_df['id'] = [row.target+'_'+row.pos for ix, row in on_df.iterrows()]\n",
    "    on_df['offexp'] = [[] for i in range(on_df.shape[0])]\n",
    "    off_df['id'] = [row.target+'_'+row.pos for ix, row in off_df.iterrows()]\n",
    "    off_df = off_df[off_df['tm'].astype(float)>arg['tm_thresh']]\n",
    "    probe_groups = off_df.groupby('id')\n",
    "    \n",
    "    counter = 0\n",
    "    for key, values in probe_groups.groups.iteritems():\n",
    "        number_offtargets_this_probe = len(values)\n",
    "        exp_values = expression_df[expression_column].ix[values]\n",
    "        agg_value = aggfunc(exp_values)\n",
    "        if len(exp_values)<1:\n",
    "            continue\n",
    "        subset = on_df[on_df['id']==key].index\n",
    "        on_df.loc[subset, 'offexp'] = [zip(off_df.loc[values, 'offtarget'].values,\n",
    "                                               exp_values.values, \n",
    "                                               off_df.loc[values, 'tm'].astype(float).values)]\n",
    "    return on_df\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
